### DISCUSSION TITLE
Segmentation fault while attempting to run Osu on a Chromebook natively

### DISCUSSION THREAD

**Discussion by JiamiWang** [ppy/osu#30449] [Q&A] [answered]
I know this is a very unique issue, and I might share this with the Chromebrew team as well, but I would like to ask for help to see if there is a way to address this, and get further insight so I could properly share this with Chromebrew.

## Description
When running the osu.AppImage file natively on the Dell Chromebook 3110, the app causes a segmentation error. This was run after installing sufficient packages (for example, Xwayland, Xorg, etc) through the Chromebrew package manager. The issue may be something involved with how Chromebrew's sommelier runner for GUI apps work, but this is not the case after using GDB (logs attached below) to validate the error that was caused.

# Steps to  reproduce this error
1) Get a Chromebook (the Dell Chromebook 3110, or perhaps one of the same `dedede` board)
2) Press ESC + Refresh + Power to get to the recovery screen
3) Press Ctrl + D on this screen to enable developer mode
4) Wait for the computer to restart
5) If this is a new Chromebook, click on Debugging mode, which will remove the rootfs write protection. I disabled it and I'm not sure if it has any effect on this error.
6) Press Ctrl + Alt + Forward page key to open the VT-2 console, and log in with chronos
7) Install Chromebrew by executing the bash script, which is found on [Chromebrew's website](https://chromebrew.github.io/), and after installing Chromebrew, this would give the ability to run GUI apps alongside a working package manager. 
8) Install Xwayland (I was lazy so I just decided to run `crew install firefox` , which installs Firefox to replace stinky Google Chrome and also the necessary packages to run GUI apps)
9) Download Osu
10) Create a new folder in `/usr/local`, and `chmod -R 777 [folder name]` that folder to grant access to all users
11) `cp` the downloaded Osu AppImage file to the folder created in `/usr/local` 
12) Run the AppImage
12a) I used --appimage-extract because `fusermount` could not mount the AppImage and then ran AppRun in the extracted files folder, not sure if this is only happening for my Chromebook
13) Voila! A segmentation fault.

**Outputs**
`uname -m`: x86_64

## Version
2024.1009.1

**Comment by JiamiWang** [comment]
Below are the GDB logs
## GDB Logs
Below is the GDB log of running osu.AppImage. Note that I did this under root because under the chronos user it would not run due to heavy sanctions imposed by the ChromeOS firmware restrictions.

```
Working directory /usr/local/apps.
warning: BFD: reopening /tmp/.mount_osu.Apuba6eD/usr/bin/osu!: Transport endpoint is not connected
Starting program: /usr/local/apps/osu.AppImage
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/local/lib64/libthread_db.so.1".
[Detaching after fork from child process 21142]
process 21139 is executing new program: /bin/dash
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/local/lib64/libthread_db.so.1".
[Detaching after fork from child process 21147]
[Detaching after vfork from child process 21152]
[Detaching after vfork from child process 21153]
[Detaching after fork from child process 21154]
[Detaching after fork from child process 21156]
process 21139 is executing new program: /tmp/.mount_osu.AplSIHML/usr/bin/osu!
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/local/lib64/libthread_db.so.1".
[New Thread 0x7ffff72e4640 (LWP 21161)]
[New Thread 0x7ffff6ad5640 (LWP 21162)]
[New Thread 0x7ffff62bf640 (LWP 21163)]
[New Thread 0x7ffff5aba640 (LWP 21164)]
[New Thread 0x7ffff51ab640 (LWP 21165)]
[New Thread 0x7ffff35a9640 (LWP 21166)]
[New Thread 0x7fbeced0d640 (LWP 21173)]
[New Thread 0x7fbece50c640 (LWP 21174)]
[New Thread 0x7fbecdd0b640 (LWP 21175)]
[New Thread 0x7fbecd50a640 (LWP 21176)]
[New Thread 0x7fffef42c640 (LWP 21177)]
[New Thread 0x7fbeccd09640 (LWP 21178)]
[New Thread 0x7fbeaffff640 (LWP 21179)]
[New Thread 0x7fbeaf064640 (LWP 21180)]
[New Thread 0x7fbeae6c7640 (LWP 21181)]
[New Thread 0x7fbe956b5640 (LWP 21182)]
[New Thread 0x7fbe94eb4640 (LWP 21183)]
[New Thread 0x7fbe946b3640 (LWP 21184)]
[New Thread 0x7fbe93eb2640 (LWP 21185)]
[New Thread 0x7fbe930b1640 (LWP 21186)]
[New Thread 0x7fbe928b0640 (LWP 21187)]
[New Thread 0x7fbe91fb9640 (LWP 21188)]
[New Thread 0x7fbe917b8640 (LWP 21189)]
[New Thread 0x7fbe90fb7640 (LWP 21190)]
[New Thread 0x7fbe907b6640 (LWP 21191)]
[New Thread 0x7fbe8ffaf640 (LWP 21192)]
[New Thread 0x7fbe88867640 (LWP 21194)]
[New Thread 0x7fbe8f628640 (LWP 21193)]
Thread 21 "Audio (GameThre" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fbe930b1640 (LWP 21186)]
0x00007ffff7b1f456 in ?? () from /usr/local/lib64/libc.so.6
```

**Reply by hwsmm** [reply]
This is not enough. Can you type `backtrace` in GDB at that point and post the output here?

**Reply by JiamiWang** [reply]
Yes, here you go:

```Thread 21 "Audio (GameThre" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fbe92fbb640 (LWP 15432)]
0x00007ffff7b1f456 in ?? () from /usr/local/lib64/libc.so.6
#0  0x00007ffff7b1f456 in ?? () from /usr/local/lib64/libc.so.6
#1  0x00007fbe8efa9005 in cras_client_get_output_devices () from /usr/local/lib64/libcras.so.0
#2  0x00007fbeac0034d3 in ctl_cras_read_integer () from /usr/local/lib64/alsa-lib/libasound_module_ctl_cras.so
#3  0x00007fbe9302f755 in ?? () from /usr/local/lib64/libasound.so
#4  0x00007fbe93040a57 in ?? () from /usr/local/lib64/libasound.so
#5  0x00007fbe930411d1 in ?? () from /usr/local/lib64/libasound.so
#6  0x00007fbe93042e52 in ?? () from /usr/local/lib64/libasound.so
#7  0x00007fbe930435b4 in ?? () from /usr/local/lib64/libasound.so
#8  0x00007fbe93043718 in ?? () from /usr/local/lib64/libasound.so
#9  0x00007fbe93031e7e in ?? () from /usr/local/lib64/libasound.so
#10 0x00007fbe9303351b in snd_hctl_load () from /usr/local/lib64/libasound.so
#11 0x00007fbe93037209 in snd_mixer_load () from /usr/local/lib64/libasound.so
#12 0x00007fbead8088d9 in ?? () from /tmp/.mount_osu.ApMW5lcK/usr/bin/libbass.so
#13 0x0000000100000001 in ?? ()
#14 0x0000000000000000 in ?? ()
warning: BFD: reopening /tmp/.mount_osu.ApMW5lcK/usr/bin/osu!: Transport endpoint is not connected
```

**Reply by JiamiWang** [reply]
The bytes of the threads will definitely not correlate to the original post because I had to run it again, but I don't think that matters too much since it is likely the same order. To make it less confusing in any case, here is the result from `run`:

```
Starting program: /usr/local/apps/osu.AppImage
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/local/lib64/libthread_db.so.1".
[Detaching after fork from child process 15392]
process 15389 is executing new program: /bin/dash
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/local/lib64/libthread_db.so.1".
[Detaching after fork from child process 15397]
[Detaching after vfork from child process 15402]
[Detaching after vfork from child process 15403]
[Detaching after fork from child process 15404]
[Detaching after fork from child process 15406]
process 15389 is executing new program: /tmp/.mount_osu.ApMW5lcK/usr/bin/osu!
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/local/lib64/libthread_db.so.1".
[New Thread 0x7ffff72e4640 (LWP 15411)]
[New Thread 0x7ffff6ad5640 (LWP 15412)]
[New Thread 0x7ffff62bf640 (LWP 15413)]
[New Thread 0x7ffff5aba640 (LWP 15414)]
[New Thread 0x7ffff51ab640 (LWP 15415)]
[New Thread 0x7ffff35a9640 (LWP 15416)]
[New Thread 0x7fbeced0d640 (LWP 15417)]
[New Thread 0x7fbece50c640 (LWP 15418)]
[New Thread 0x7fbecdd0b640 (LWP 15419)]
[New Thread 0x7fbecd50a640 (LWP 15420)]
[New Thread 0x7fffef42c640 (LWP 15421)]
[New Thread 0x7fbeccd09640 (LWP 15422)]
[New Thread 0x7fbeaffff640 (LWP 15423)]
[New Thread 0x7fbeaef9f640 (LWP 15424)]
[New Thread 0x7fbeae6c7640 (LWP 15425)]
[New Thread 0x7fbe956b5640 (LWP 15427)]
[New Thread 0x7fbe94eb4640 (LWP 15428)]
[New Thread 0x7fbe946b3640 (LWP 15429)]
[New Thread 0x7fbe93eb2640 (LWP 15430)]
[New Thread 0x7fbe92fbb640 (LWP 15432)]
[New Thread 0x7fbe927ba640 (LWP 15433)]
[New Thread 0x7fbe91fb9640 (LWP 15434)]
[New Thread 0x7fbe917b8640 (LWP 15435)]
[New Thread 0x7fbe90fb7640 (LWP 15436)]
[New Thread 0x7fbe907b6640 (LWP 15437)]
[New Thread 0x7fbe8ffb5640 (LWP 15438)]
[New Thread 0x7fbe8f7b4640 (LWP 15439)]
[New Thread 0x7fbe88908640 (LWP 15440)]

Thread 21 "Audio (GameThre" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fbe92fbb640 (LWP 15432)]
0x00007ffff7b1f456 in ?? () from /usr/local/lib64/libc.so.6
```

**Reply by hwsmm** [reply]
It sounds like a bug in libcras, which is an audio server for ChromeOS (and I've never heard of it until now).

But I have heard a few times that osu!lazer works on ChromeOS with some workarounds. Search ChromeOS in this discussion section and maybe you can ask them about the steps you need to take to get it running.

**Reply by JiamiWang** [reply, marked as answer]
Well, none of the other posts are specifically about running Osu native on ChromeOS, but rather Chrostini, a Linux virtual machine which will be very laggy when ran on my device since it has poor hardware. I remember I booted into ChromeOS unprovisioned using [shimboot](https://github.com/ading2210/shimboot) which opened Osu but did not have sound support for Dedede, so no sound cards was available and it worked fine. I think I will just share this with Chromebrew and see what they can do about this.
(edit 1) Or maybe its just that sound support for Dedede is just poor in general I guess
